farmerbot/server/farmerbot.go

```diff
@@ -0,0 +1,296 @@
+		if err != nil {
+			log.Error().Err(err).Msg("[FARMERBOT] failed to perform periodic wake up")
+		}
+
+		log.Debug().Msg("[FARMERBOT] power management")
+		err = f.powerManager.PowerManagement()
+		if err != nil {
+			log.Error().Err(err).Msg("[FARMERBOT] failed to power management nodes")
+		}
+
+		delta := time.Since(startTime)
+		log.Debug().Float64("Elapsed time for updates in minutes", delta.Minutes())
+
+		// sleep if finished before the update timeout
+		var timeToSleep float64
+		if delta.Minutes() >= constants.TimeoutUpdate.Minutes() {
+			timeToSleep = 0
+		} else {
+			timeToSleep = constants.TimeoutUpdate.Minutes() - delta.Minutes()
+		}
+		time.Sleep(time.Duration(timeToSleep) * time.Minute)
+	}
+}
+
+func (f *FarmerBot) serve(ctx context.Context) {
+	router := peer.NewRouter()
+	farmerbot := router.SubRoute("farmerbot")
+
+	farmRouter := farmerbot.SubRoute("farmmanager")
+	nodeRouter := farmerbot.SubRoute("nodemanager")
+	powerRouter := farmerbot.SubRoute("powermanager")
+
+	// TODO: didn't work
+	// powerRouter.Use(f.authorize)
+
+	farmRouter.WithHandler("version", func(ctx context.Context, payload []byte) (interface{}, error) {
+		return version.Version, nil
+	})
+
+	nodeRouter.WithHandler("findnode", func(ctx context.Context, payload []byte) (interface{}, error) {
+		var options models.NodeOptions
+
+		if err := json.Unmarshal(payload, &options); err != nil {
+			return nil, fmt.Errorf("failed to load request payload: %w", err)
+		}
+
+		nodeID, err := f.powerManager.FindNode(options)
+		return nodeID, err
+	})
+
+	powerRouter.WithHandler("poweroff", func(ctx context.Context, payload []byte) (interface{}, error) {
+		err := f.authorize(ctx)
+		if err != nil {
+			return nil, err
+		}
+
+		var nodeID uint32
+		if err := validateAccountEnoughBalance(f.identity, f.sub); err != nil {
+			return nil, fmt.Errorf("failed to validate account balance: %w", err)
+		}
+
+		if err := json.Unmarshal(payload, &nodeID); err != nil {
+			return nil, fmt.Errorf("failed to load request payload: %w", err)
+		}
+
+		return nil, f.powerManager.PowerOff(nodeID)
+	})
+
+	powerRouter.WithHandler("poweron", func(ctx context.Context, payload []byte) (interface{}, error) {
+		err := f.authorize(ctx)
+		if err != nil {
+			return nil, err
+		}
+
+		var nodeID uint32
+		if err := validateAccountEnoughBalance(f.identity, f.sub); err != nil {
+			return nil, fmt.Errorf("failed to validate account balance: %w", err)
+		}
+
+		if err := json.Unmarshal(payload, &nodeID); err != nil {
+			return nil, fmt.Errorf("failed to load request payload: %w", err)
+		}
+
+		err = f.powerManager.PowerOn(nodeID)
+		return nil, err
+	})
+
+	_, err := peer.NewPeer(
+		ctx,
+		peer.KeyTypeSr25519,
+		f.mnemonic,
+		constants.RelayURLS[f.network],
+		fmt.Sprintf("farmerbot-%d", f.config.Farm.ID),
+		&f.sub,
+		false,
+		router.Serve,
+	)
+
+	if err != nil {
+		log.Fatal().Err(err).Msg("[FARMERBOT] failed to create farmerbot direct peer")
+	}
+
+	select {}
```

Comment:
```suggestion
	<-ctx.Done()
```

we use `select{}` when we want to block forever with no chance of unlocking, but we already have a context that can be cancelled then `<-ctx.Done()` makes more sense , right ?

**not right**. Because this method (in this specific use case) does not need to be blocked, in the matter of fact you don't even need to start it in a `go routine`. I mean think about it, why start it in the background then block it from returing.

Creating the Peer makes it start serving requests immedilatey already in the background. Which means this method itself can return (also you can return an error)

 I used `select{}` in the example code because there was only the main routing and there was nothing that can stop the application from exiting completely, unlike here. So it's important to understand why things are done and change it to work with our usecase.